app/src/util/env_config.py

```diff
@@ -1,16 +1,30 @@
-import os
+import logging
 
-from pydantic import BaseSettings
+from pydantic import BaseModel
 
-import src
+logger = logging.getLogger(__name__)
 
-env_file = os.path.join(
-    os.path.dirname(os.path.dirname(src.__file__)),
-    "config",
-    "%s.env" % os.getenv("ENVIRONMENT", "local"),
-)
 
+class PydanticBaseEnvConfig(BaseModel):
+    """Base class for application configuration.
+
+    Similar to Pydantic's BaseSettings class, but we implement our own method to override from the
+    environment so that it can be run later, after an instance was constructed."""
 
-class PydanticBaseEnvConfig(BaseSettings):
     class Config:
-        env_file = env_file
+        validate_assignment = True
+
+    def override_from_environment(self, environ, prefix=""):
+        """Recursively override field values from the given environment variable mapping."""
+        for name, field in self.__fields__.items():
+            if field.is_complex():
+                # Nested models must be instances of this class too.
+                getattr(self, name).override_from_environment(environ, prefix=name + "_")
+                continue
+
+            env_var_name = field.field_info.extra.get("env", prefix + name)
+            for key in (env_var_name, env_var_name.lower(), env_var_name.upper()):
```

Comment:
Might be nice to be stricter here and require a certain casing, avoid letting someone accidentally set the environment variable in lowercase and then not understand why it's not being overridden if it's also set in uppercase elsewhere.